Sources/Extraction/Filters/HillOrientation.c

```diff
@@ -1,26 +1,183 @@
 #include <stdlib.h>
 #include <assert.h>
+#include <math.h>
+#include <stdio.h>
 
 #include "HillOrientation.h"
+#include "General/Array.h"
 
-HillOrientation HillOrientation_Construct(void)
+#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))
+#define MAX(X, Y) (((X) > (Y)) ? (X) : (Y))
+#define NUM_VECTORS (28)
+
+const Point neighbors[NUM_VECTORS] = {
+    { -1, 1, }, { -1, 2 }, { -1, 3 },
+    { -2, 1  }, { -2, 2 }, { -2, 3 },
+    { -3, 1  }, { -3, 2 }, { -3, 3 },
+    { 0, 1 }, { 0, 2, }, { 0, 3 },
+    { 1, 0 }, { 1, 1, }, { 1, 2 }, { 1, 3 },
+    { 2, 0 }, { 2, 1  }, { 2, 2 }, { 2, 3 },
+    { 3, 0 }, { 3, 1  }, { 3, 2 }, { 3, 3 },
+    { -5, 0 }, { -5, 5 }, { 0, 5 }, { 5, 5 }
+};
+
+static bool IsInsideImageAndMask(int imageHeight, int imageWidth, BoolArray2D pixelMask, int x, int y)
+{
+    bool ok = x >= 0
+        && y >= 0
+        && x < imageWidth
+        && y < imageHeight
+        && pixelMask.data[x][y];
+
+    return ok;
+}
+
+/*
+static void print_directions(PointFArray2D directions) {
+  for ( int i=0; i < directions.sizeX; ++i ) {
+    for ( int j=0; j < directions.sizeY; ++j ) {
+      printf("(%f, %f)", directions.data[i][j].x, directions.data[i][j].y);
+    }
+    printf("\n");
+  }
+}
+*/
+
+/*
+static void print_BoolArray2D(BoolArray2D array) {
+  for ( int i=0; i < array.sizeX; ++i ) {
+    for ( int j=0; j < array.sizeY; ++j ) {
+      printf("%d ", array.data[i][j]);
+    }
+    printf("\n");
+  }
+} */
+
+PointFArray2D HillOrientation_AccumulateDirections(FloatArray2D input, int imageHeight, int imageWidth, BoolArray2D pixelMask)
```

Comment:
Why not pass in Size to start off with and then you don;t need to re-create the Size object thingy